### ISSUE TITLE
Docker pull progressively slow for large layers

### ISSUE THREAD

**Issue by vronin-okta** [docker/docker#13309]
I am pulling an image from a private docker repo. It consist from multiple layers. I can see how small layers (as example 20MB) downloaded in 5-6 second (~4Mb/s).

However, with bigger layers (as example 300MB) it works fast at the beginning, but starts to slow down at the end and gets to ~0.1Mb/s (40 times slower than smaller layers or beginning of this layer)

**Comment by thaJeztah** [comment]
Can you add the requested information (see above)?

**Comment by vronin-okta** [comment]
Here is the info:
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2

`docker info`:
Containers: 3
Images: 62
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 68
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-24-generic
Operating System: Ubuntu 14.04 LTS
CPUs: 4
Total Memory: 7.798 GiB
Name: ubuntu-14
ID: XPCI:UBOH:JE7H:QB5G:MHO6:75HZ:KZOB:QJZF:ZBTN:N5MO:5CKY:3NMZ

`uname -a`
Linux ubuntu-14 3.13.0-24-generic #46-Ubuntu SMP Thu Apr 10 19:11:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

Environment:
VirtualBox (managed by Vagrant)

Steps to reproduce:
1) I am using Artifactory as a docker registry
2) I am trying to do docker pull
3) It starts dowloading layers and I can see that small layers are downloaded fast, but big layers are downloaded slow

I expected the speed to be approximately the same (don't wary tenfold or twentyfold).

**Comment by vronin-okta** [comment]
I understand that my issue is quite vague. However, I was trying to find info in internet about docker pull slowness and on one hand a lot of people says that it exist on other hand, there were couple of bugs here related to it which are marked as closed.

I was able to download ubuntu docker image which was 0.2Gb without any problem (really fast). And my custom image which was 1.6Gb took forever (something like half an hour).

I checked that it's not my connection speed.
a) I used speed test
b) I downloaded other artifacts from Artifactory and the speed was really high

I understand that there is one more variable here, which is Artifactory. It could be that it's involved. I don't have another private docker registry to try it out.

**Comment by jberkus** [comment]
I've had the same experience with the Docker Hub, both pushing and pulling.  For example, I recently uploaded an image to the hub which had layers from 15Mb to 200MB.  The 15MB layers uploaded at around 1MB/s, but the 200MB layer uploaded at 0.1MB/s, causing it to take over half an hour.  I'd assume that this was just the vaguearies of the internet, except that the pattern seemed really consistent, and I've seen lots of similar reports on Stack Overflow etc.

Is Docker maybe doing some kind of progressive checksumming as the file is downloaded/uploaded?  Something else which would involve doing a cyclical check on the contents of the whole file, and thus be geometrically slower with large files?

**Comment by iadknet** [comment]
I have also been struggling with pull performance from a private registry.  As far as I can tell, I think it is getting stuck on "verifying checksum"   But pulling, what should just be a couple of ~100MB layers takes close to 10 minutes.

**Comment by stevvooe** [comment]
From #14018:

> I've been able to reproduce that issue multiple times; it really seems like docker push checksums each MB as it's uploaded, resulting in N^2 upload times for large layers.

I took a quick peak at this and it looks like there is a performance knee for large files, rather than a particular N^2 process. If we look at https://github.com/docker/docker/blob/master/graph/push_v2.go#L225, we can see that the file is buffered into a temp file, then read back. This is a 4*N process. Mathematically, this is okay, but this is probably causing the issue. On small files, this is probably done mostly in buffer cache and it doesn't have an effect on fast disks. For larger files with fast disks, this is going to have a massive impact.

I'm putting together a PR to fix this now.

**Comment by jberkus** [comment]
@stevvooe Sure I could see that.  I don't have a good way to systematically test pushes of each different size to determine the exact pattern.  

Let me know when/where I can test to see if you've fixed it.

**Comment by thaJeztah** [comment]
@jberkus PR is up here: https://github.com/docker/docker/issues/15493, so you could try to build from that

**Comment by jberkus** [comment]
OK, so this fix is entirely client-side?  I assumed that there needed to be a fix on the Hub side as well.

**Comment by stevvooe** [comment]
@jberkus Please avoid assumptions and let's focus on facts and real observations.

#14018 is covering _push_ performance issues. From the looks of this issue, it is not covering _push_ performance. My comments earlier and #15493 cover _pull_ performance. Sorry for the confusion (I made assumptions based on the extra comments).

@vronin-okta We need some more information to troubleshoot the original complaint. Would you mind providing the docker logs with debug enabled on pull? We may also need to get someone from sonatype involved to help debug the registry side.

@jberkus I am not sure that you're particular issue is covered here. I would consider this bug to be artifactory + pull, whereas your issue is hub + pull/push. These may have shared causes but could be wildly different.

**Comment by jberkus** [comment]
@stevvooe ooops, sorry. too many pull/push issues open that I'm trying to test.

**Comment by stevvooe** [comment]
@jberkus No worries. We are trying to keep this manageable ourselves.

Did you have a chance to try out #15493? I think this one addresses the perceived scaling behavior? Others have had luck with it.

**Comment by jberkus** [comment]
@stevvooe : I'm having trouble finding the final version of the patch on #15493.  It's a bit of a mess.

**Comment by stevvooe** [comment]
@jberkus There is not a version with that integrated. You'll have to build it yourself. Its up to you whether you want to try now or wait till it is merged and you can give master a try, which is always available at https://master.dockerproject.org/.

We won't close this till we get some confirmation the worst causes of this issue are addressed.

**Comment by jberkus** [comment]
@stevvooe: oh, looks like I can just pull it from your fork.  What I was questioning is whether it was ready for testing; there's multiple code comments on that PR, and it wasn't clear that they were all resolved.

**Comment by stevvooe** [comment]
@jberkus Don't worry about those. Most are just nits and sloppy integration tests.

**Comment by jberkus** [comment]
OK.  This'll take a bit; the system where I have images with large layers is not the same system where I have docker built from source.  Probably I should set up docker-on-docker for source builds ...

**Comment by stevvooe** [comment]
@jberkus We merged #15493 into master. Could you give it a try? Binaries are available here: https://master.dockerproject.org/.

**Comment by jberkus** [comment]
Will do!

**Comment by tiborvass** [comment]
@jberkus any updates?

**Comment by jberkus** [comment]
Finally got a test environment up.  The issue appears to be fixed in the latest; at this point, pull speed seems to be constant with the size of the layer.

**Comment by stevvooe** [comment]
@jberkus Great to hear!

**Comment by owendall** [comment]
I am having the slowest issue trying to pull an image for Oracle.  See attached, which was started yesterday, got 401 errors, and now started again.  Any help will be appreciated
![docker pull slow and errored out]

Just installed the Docker tools yesterday on my Mac, and am connected via ethernet to Verizon Fios.

(https://cloud.githubusercontent.com/assets/10927/11918892/7985d3da-a70f-11e5-9342-fe2f857b5330.png)

**Comment by stevvooe** [comment]
@owendall Please do not comment on closed issues. It is extremely unlikely that you are encountering this specific problem and the data you've provided does not indicate that in any way.

**Comment by Milewski** [comment]
for me (in china) download a 8 mb directly from an high end server (something closer to amazon Chinese made).. takes like forever....

**Comment by giovannicandido** [comment]
I have this experience in every system with docker installed, OSX, Linux, PhantonOS etc..
Pulling images from the registry starts quickly and then drops to really slow download.

Is docker pulling layers in parallel? 
I think the issue is related to parallel image download from docker. Something in the network drops or slow down the connection (maybe the firewall server from docker hub registry) because it think is  a abuse too many connection from the same origin.

I'm working on a 800Mbs link with 1Gb switchs, my machine can't even saturate the link because my Hard Drive is slower lolol, so is not a network issue, and I have tested that possibility

**Comment by jberkus** [comment]
This issue is closed, as the original problem was resolved.  If you think you've discovered a new issue with pulls, please open a new issue.  Thanks!

**Comment by thaJeztah** [comment]
> Is docker pulling layers in parallel? I guess the issue is related to parallel image download from docker. 

Yes, image layers are pushed/pulled in parallel. You can set the number of concurrent up/downloads in the daemon configuration, using a flag;

```
     --max-concurrent-downloads=3           Set the max concurrent downloads for each pull
      --max-concurrent-uploads=5             Set the max concurrent uploads for each push
```

Or through the `daemon.json` configuration file;
https://github.com/docker/docker/blob/v1.12.3/docs/reference/commandline/dockerd.md#linux-configuration-file

Docker for Mac (at least in the Beta, not sure in the "stable" channel), allows setting this configuration in the UI;
<img width="380" alt="screen shot 2016-12-01 at 18 42 04" src="https://cloud.githubusercontent.com/assets/1804568/20805170/f756e252-b7f5-11e6-8e88-dbc0275d2b6f.png">

I hope this information helps you further, I'm going to lock the discussion here, because (as @jberkus explained), closed issues should not be used for reporting new issues. It's easy for reports to get unnoticed when commenting on closed issues, and if the issue is unrelated, makes it hard to keep track.
